libsolidity/ast/AST.cpp

```diff
@@ -641,86 +615,62 @@ FunctionTypePointer VariableDeclaration::functionType(bool _internal) const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<VariableDeclarationAnnotation>();
-	return dynamic_cast<VariableDeclarationAnnotation&>(*m_annotation);
+	return initAnnotation<VariableDeclarationAnnotation>();
 }
 
 StatementAnnotation& Statement::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<StatementAnnotation>();
-	return dynamic_cast<StatementAnnotation&>(*m_annotation);
+	return initAnnotation<StatementAnnotation>();
 }
 
 InlineAssemblyAnnotation& InlineAssembly::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<InlineAssemblyAnnotation>();
-	return dynamic_cast<InlineAssemblyAnnotation&>(*m_annotation);
+	return initAnnotation<InlineAssemblyAnnotation>();
 }
 
 BlockAnnotation& Block::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<BlockAnnotation>();
-	return dynamic_cast<BlockAnnotation&>(*m_annotation);
+	return initAnnotation<BlockAnnotation>();
 }
 
 TryCatchClauseAnnotation& TryCatchClause::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<TryCatchClauseAnnotation>();
-	return dynamic_cast<TryCatchClauseAnnotation&>(*m_annotation);
+	return initAnnotation<TryCatchClauseAnnotation>();
 }
 
 ForStatementAnnotation& ForStatement::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<ForStatementAnnotation>();
-	return dynamic_cast<ForStatementAnnotation&>(*m_annotation);
+	return initAnnotation<ForStatementAnnotation>();
 }
 
 ReturnAnnotation& Return::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<ReturnAnnotation>();
-	return dynamic_cast<ReturnAnnotation&>(*m_annotation);
+	return initAnnotation<ReturnAnnotation>();
 }
 
 ExpressionAnnotation& Expression::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<ExpressionAnnotation>();
-	return dynamic_cast<ExpressionAnnotation&>(*m_annotation);
+	return initAnnotation<ExpressionAnnotation>();
 }
 
 MemberAccessAnnotation& MemberAccess::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<MemberAccessAnnotation>();
-	return dynamic_cast<MemberAccessAnnotation&>(*m_annotation);
+	return initAnnotation<MemberAccessAnnotation>();
 }
 
 BinaryOperationAnnotation& BinaryOperation::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<BinaryOperationAnnotation>();
-	return dynamic_cast<BinaryOperationAnnotation&>(*m_annotation);
+	return initAnnotation<BinaryOperationAnnotation>();
 }
 
 FunctionCallAnnotation& FunctionCall::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<FunctionCallAnnotation>();
-	return dynamic_cast<FunctionCallAnnotation&>(*m_annotation);
+	return initAnnotation<FunctionCallAnnotation>();
 }
 
 IdentifierAnnotation& Identifier::annotation() const
 {
-	if (!m_annotation)
-		m_annotation = make_unique<IdentifierAnnotation>();
-	return dynamic_cast<IdentifierAnnotation&>(*m_annotation);
+	return initAnnotation<IdentifierAnnotation>();
```

Comment:
Couldn't the entire `annotation()` implementation moved to a template if all of them just call this anyway?

Reply:
My thought, too, but I think it's fine to merge like this for now.